src/textual/widgets/_input.py

```diff
@@ -29,6 +30,13 @@
 """Set literal with the legal values for the type `InputValidationOn`."""
 
 
+_RESTRICT_TYPES = {
+    "integer": r"[-+]?\d*",
+    "number": r"[-+]?\d*\.?\d*[eE]?-?\d*",
```

Comment:
Looking at this, and in a quick test, this doesn't seem to allow for something such as `1e+1`.

```python
>>> float("1e+1")
10.0
```